File: ieee2022/java/src/DisjointSubsets/DisjointSubsets.java
package DisjointSubsets;

import java.util.AbstractMap.SimpleImmutableEntry;
import java.util.HashMap;
import java.util.Optional;

public class DisjointSubsets<T> {

    public HashMap<T, UpTreeSize<T>> elements;

    public DisjointSubsets() {
        this.elements = new HashMap<>();
    }

    public void add(T index) {
        this.elements.put(index, new UpTreeSize<T>(index));
    }

    /*public void addWithCheck(T index, K data) {
        if (!this.elements.containsKey(index)) {
            this.elements.put(index, new SimpleImmutableEntry<>(new UpTreeSize<T>(index), data));
        }
    }*/

    public UpTreeSize<T> find(T index) {
        return this.elements.get(index);
    }

    /*public Optional<SimpleImmutableEntry<UpTreeSize<T>, K>> optionalFind(T index) {
        return Optional.ofNullable(this.elements.get(index));
    }

    public SimpleImmutableEntry<UpTreeSize<T>, K> findWithAdd(T index, K defaultData) {
        if (this.elements.containsKey(index)) {
            return this.elements.get(index);
        }
        this.elements.put(index, new SimpleImmutableEntry<>(new UpTreeSize<T>(index), defaultData));
        return this.elements.get(index);
    }*/

    public void join(T a, T b) {
        UpTreeSize.union(this.find(a), this.find(b));
    }

    /*public boolean joinWithCheck(T a, T b) {
        Optional<SimpleImmutableEntry<UpTreeSize<T>, K>> aElement = this.optionalFind(a);
        if (aElement.isEmpty()) {
            return false;
        }
        Optional<SimpleImmutableEntry<UpTreeSize<T>, K>> bElement = this.optionalFind(b);
        if (bElement.isEmpty()) {
            return false;
        }
        UpTreeSize.union(aElement.get().getKey(), bElement.get().getKey());
        return true;
    }

    public void joinWithAdd(T a, T b, K aDefaultData, K bDefaultData) {
        SimpleImmutableEntry<UpTreeSize<T>, K> aElement = this.findWithAdd(a, aDefaultData);
        SimpleImmutableEntry<UpTreeSize<T>, K> bElement = this.findWithAdd(b, bDefaultData);
        UpTreeSize.union(aElement.getKey(), bElement.getKey());
    }*/

    public boolean inSame(T a, T b) {
        return UpTreeSize.inSameSet(this.find(a), this.find(b));
    }

    /*public boolean inSameWithCheck(T a, T b) {
        Optional<SimpleImmutableEntry<UpTreeSize<T>, K>> aElement = this.optionalFind(a);
        if (aElement.isEmpty()) {
            return false;
        }
        Optional<SimpleImmutableEntry<UpTreeSize<T>, K>> bElement = this.optionalFind(b);
        if (bElement.isEmpty()) {
            return false;
        }
        return UpTreeSize.inSameSet(aElement.get().getKey(), bElement.get().getKey());
    }*/
}

File: ieee2022/python/disjoint_subsets/disjoint_subsets.py

from up_tree_order import UpTreeOrder as UpTree
# from up_tree_size import UpTreeSize as UpTree




class DisjointSubsets:
    def __init__(self):
        self.elements = {}

    def add(self, index, data):
        self.elements[index] = (UpTree(index), data)

    def add_with_check(self, index, data):
        if index not in self.elements.keys():
            self.elements[index] = (UpTree(index), data)

    def find(self, index):
        return self.elements[index]

    def find_with_check(self, index):
        if index in self.elements.keys():
            return self.elements[index]

    def find_with_add(self, index, default_data):
        if index in self.elements.keys():
            return self.elements[index]
        self.elements[index] = (UpTree(index), default_data)
        return self.elements[index]

    def join(self, a, b):
        UpTree.union(self.find(a)[0], self.find(b)[0])

    def join_with_check(self, a, b):
        a_el = self.find_with_check(a)
        if a_el is None:
            return False
        b_el = self.find_with_check(b)
        if b_el is None:
            return False
        UpTree.union(a_el[0], b_el[0])
        return True

    def join_with_add(self, a, b, default_data_a, default_data_b):
        a_el = self.find_with_add(a, default_data_a)
        b_el = self.find_with_add(b, default_data_b)
        UpTree.union(a_el[0], b_el[0])

    def in_same(self, a, b):
        a_el = self.find(a)
        b_el = self.find(b)
        return UpTree.in_same_set(a_el[0], b_el[0])

    def in_same_with_check(self, a, b):
        a_el = self.find_with_check(a)
        if a_el is None:
            return False
        b_el = self.find_with_check(b)
        if b_el is None:
            return False
        return UpTree.in_same_set(a_el[0], b_el[0])
